File: mobile/lib/services/notification_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:http/http.dart' as http;
import '../utils/ip_detection.dart';

class NotificationService {
  static final FirebaseMessaging _firebaseMessaging =
      FirebaseMessaging.instance;
  static final FlutterLocalNotificationsPlugin _localNotifications =
      FlutterLocalNotificationsPlugin();

  static String? _fcmToken;
  static bool _isInitialized = false;

  /// Inicializar el servicio de notificaciones
  static Future<void> initialize() async {
    if (_isInitialized) return;

    try {
      // Configurar notificaciones locales
      await _setupLocalNotifications();

      // Configurar Firebase Messaging
      await _setupFirebaseMessaging();

      // Configurar listeners
      _setupMessageHandlers();

      _isInitialized = true;
    } catch (e) {
      rethrow;
    }
  }

  /// Registrar dispositivo después del login exitoso
  static Future<void> registerDeviceAfterLogin() async {
    await _registerDeviceWithBackend();
  }

  /// Configurar notificaciones locales
  static Future<void> _setupLocalNotifications() async {
    const AndroidInitializationSettings initializationSettingsAndroid =
        AndroidInitializationSettings('@mipmap/ic_launcher');

    const DarwinInitializationSettings initializationSettingsDarwin =
        DarwinInitializationSettings(
          requestAlertPermission: true,
          requestBadgePermission: true,
          requestSoundPermission: true,
        );

    const InitializationSettings initializationSettings =
        InitializationSettings(
          android: initializationSettingsAndroid,
          iOS: initializationSettingsDarwin,
        );

    await _localNotifications.initialize(
      initializationSettings,
      onDidReceiveNotificationResponse: _onNotificationTapped,
    );

    // Configurar canal para Android
    if (Platform.isAndroid) {
      await _createNotificationChannel();
    }
  }

  /// Crear canal de notificación para Android
  static Future<void> _createNotificationChannel() async {
    const AndroidNotificationChannel channel = AndroidNotificationChannel(
      'transport_notifications', // ID del canal
      'Notificaciones de Transporte', // Nombre
      description:
          'Notificaciones sobre viajes, reservas y actualizaciones del sistema',
      importance: Importance.high,
      playSound: true,
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<
          AndroidFlutterLocalNotificationsPlugin
        >()
        ?.createNotificationChannel(channel);
  }

  /// Configurar Firebase Messaging
  static Future<void> _setupFirebaseMessaging() async {
    // Solicitar permisos
    NotificationSettings settings = await _firebaseMessaging.requestPermission(
      alert: true,
      announcement: false,
      badge: true,
      carPlay: false,
      criticalAlert: false,
      provisional: false,
      sound: true,
    );

    if (settings.authorizationStatus != AuthorizationStatus.authorized) {
      return;
    }

    // Obtener token FCM
    _fcmToken = await _firebaseMessaging.getToken();
  }

  /// Configurar manejadores de mensajes
  static void _setupMessageHandlers() {
    // Mensaje recibido cuando la app está en primer plano
    FirebaseMessaging.onMessage.listen(_handleForegroundMessage);

    // Mensaje tocado cuando la app está en background
    FirebaseMessaging.onMessageOpenedApp.listen(_handleMessageOpenedApp);

    // Mensaje recibido cuando la app está terminada (requiere configuración adicional)
    _handleInitialMessage();

    // Listener para actualización de token
    _firebaseMessaging.onTokenRefresh.listen((newToken) {
      _fcmToken = newToken;
      _registerDeviceWithBackend();
    });
  }

  /// Manejar mensaje inicial (app abierta desde notificación)
  static Future<void> _handleInitialMessage() async {
    RemoteMessage? initialMessage = await _firebaseMessaging
        .getInitialMessage();
    if (initialMessage != null) {
      _handleMessageOpenedApp(initialMessage);
    }
  }

  /// Manejar mensaje en primer plano
  static Future<void> _handleForegroundMessage(RemoteMessage message) async {
    // Mostrar notificación local
    await _showLocalNotification(
      title: message.notification?.title ?? 'Nueva notificación',
      body: message.notification?.body ?? 'Tienes una nueva notificación',
      payload: jsonEncode(message.data),
    );
  }

  /// Mostrar notificación local
  static Future<void> _showLocalNotification({
    required String title,
    required String body,
    String? payload,
  }) async {
    const AndroidNotificationDetails androidDetails =
        AndroidNotificationDetails(
          'transport_notifications',
          'Notificaciones de Transporte',
          channelDescription: 'Notificaciones sobre viajes y reservas',
          importance: Importance.high,
          priority: Priority.high,
          showWhen: true,
          icon: '@mipmap/ic_launcher',
        );

    const DarwinNotificationDetails iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const NotificationDetails platformDetails = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    int id = DateTime.now().millisecondsSinceEpoch ~/ 1000;

    await _localNotifications.show(
      id,
      title,
      body,
      platformDetails,
      payload: payload,
    );
  }

  /// Manejar mensaje cuando la app se abre desde notificación
  static Future<void> _handleMessageOpenedApp(RemoteMessage message) async {
    // Marcar notificación como entregada
    await _markNotificationAsDelivered(message);

    // Navegar según el tipo de notificación
    _handleNotificationNavigation(message);
  }

  /// Manejar toque en notificación local
  static Future<void> _onNotificationTapped(
    NotificationResponse response,
  ) async {
    if (response.payload != null) {
      try {
        final data = jsonDecode(response.payload!);

        // Crear RemoteMessage simulado para manejar navegación
        final message = RemoteMessage(data: Map<String, String>.from(data));

        _handleNotificationNavigation(message);
      } catch (e) {
        // Ignorar errores de parseo
      }
    }
  }

  /// Registrar dispositivo en el backend
  static Future<void> _registerDeviceWithBackend() async {
    if (_fcmToken == null) {
      return;
    }

    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('auth_token');

      if (token == null) {
        return;
      }

      final baseUrl = await IPDetection.getBaseUrl();
      await http.post(
        Uri.parse('$baseUrl/api/notificaciones/dispositivos/'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'token_fcm': _fcmToken,
          'tipo_dispositivo': Platform.isIOS ? 'ios' : 'android',
          'nombre_dispositivo': '${Platform.operatingSystem} Device',
        }),
      );
    } catch (e) {
      // Ignorar errores de registro
    }
  }

  /// Marcar notificación como entregada
  static Future<void> _markNotificationAsDelivered(
    RemoteMessage message,
  ) async {
    final notificationId = message.data['notificacion_id'];
    if (notificationId == null) return;

    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('auth_token');

      if (token == null) return;

      final baseUrl = await IPDetection.getBaseUrl();
      await http.patch(
        Uri.parse(
          '$baseUrl/api/notificaciones/notificaciones/$notificationId/',
        ),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({'fue_entregada': true}),
      );
    } catch (e) {
      // Ignorar errores al marcar notificación
    }
  }

  /// Manejar navegación por notificación
  static void _handleNotificationNavigation(RemoteMessage message) {
    final tipo = message.data['tipo'] ?? message.data['modulo'] ?? 'general';

    // Aquí puedes implementar la lógica de navegación específica
    // usando tu router existente
    switch (tipo) {
      case 'nuevo_viaje':
        // Navegar a lista de viajes disponibles
        break;
      case 'viaje_cancelado':
        // Navegar a detalles del viaje o mis reservas
        break;
      case 'recordatorio':
        // Navegar a detalles del viaje
        break;
      default:
        break;
    }
  }

  /// Obtener token FCM actual
  static String? get fcmToken => _fcmToken;

  /// Verificar si las notificaciones están habilitadas
  static Future<bool> areNotificationsEnabled() async {
    final settings = await _firebaseMessaging.getNotificationSettings();
    return settings.authorizationStatus == AuthorizationStatus.authorized;
  }

  /// Solicitar permisos de notificaciones
  static Future<bool> requestNotificationPermissions() async {
    final settings = await _firebaseMessaging.requestPermission();
    return settings.authorizationStatus == AuthorizationStatus.authorized;
  }

  /// Desactivar dispositivo en el backend
  static Future<void> unregisterDevice() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final token = prefs.getString('auth_token');

      if (token == null || _fcmToken == null) return;

      // Obtener ID del dispositivo (necesitarías implementar un endpoint para esto)
      // Por ahora, simplemente enviamos el token para que el backend lo desactive
    } catch (e) {
      // Ignorar errores al desregistrar
    }
  }
}

File: frontend/src/services/notificationService.ts
import { messaging, getToken, onMessage } from "../lib/firebase";
import { getApiBaseUrl } from "../lib/api";

export interface NotificationData {
  id: string;
  titulo: string;
  mensaje: string;
  tipo: string;
  prioridad: "baja" | "normal" | "alta" | "critica";
  data_extra?: Record<string, any>;
  fecha_creacion: string;
  fecha_lectura: string | null;
  fue_leida?: boolean; // Computed field
}

export interface NotificationPreferences {
  nuevos_viajes: boolean;
  cambios_viaje: boolean;
  recordatorios: boolean;
  promociones: boolean;
  sistema: boolean;
  push_web: boolean;
  no_molestar_desde?: string;
  no_molestar_hasta?: string;
}

class NotificationService {
  private static instance: NotificationService;
  private isInitialized = false;
  private currentToken: string | null = null;
  private onMessageCallback?: (payload: any) => void;

  public static getInstance(): NotificationService {
    if (!NotificationService.instance) {
      NotificationService.instance = new NotificationService();
    }
    return NotificationService.instance;
  }

  /**
   * Inicializar el servicio de notificaciones web
   */
  public async initialize(): Promise<void> {
    if (this.isInitialized) return;

    try {
      // Verificar si el navegador soporta notificaciones
      if (!("Notification" in window)) {
        console.warn("Este navegador no soporta notificaciones");
        return;
      }

      // Verificar si Service Worker está disponible
      if (!("serviceWorker" in navigator)) {
        console.warn("Service Worker no está disponible");
        return;
      }

      // Registrar service worker para Firebase Messaging
      await this.registerServiceWorker();

      // Configurar listeners de mensajes
      this.setupMessageListeners();

      this.isInitialized = true;
    } catch (error) {
      console.error("Error al inicializar NotificationService:", error);
      throw error;
    }
  }

  /**
   * Registrar Service Worker
   */
  private async registerServiceWorker(): Promise<void> {
    try {
      const registration = await navigator.serviceWorker.register(
        "/firebase-messaging-sw.js"
      );
      console.log("Service Worker registrado:", registration);
    } catch (error) {
      console.error("Error al registrar Service Worker:", error);
    }
  }

  /**
   * Solicitar permisos de notificación y obtener token FCM
   */
  public async requestPermissionAndGetToken(): Promise<string | null> {
    try {
      // Solicitar permisos
      const permission = await Notification.requestPermission();

      if (permission !== "granted") {
        console.warn("Permisos de notificación denegados");
        return null;
      }

      // Obtener token FCM
      const token = await getToken(messaging, {
        vapidKey: import.meta.env.VITE_FIREBASE_VAPID_KEY || "tu-vapid-key",
      });

      if (token) {
        console.log("Token FCM obtenido:", token);
        this.currentToken = token;

        // Registrar token en el backend
        await this.registerTokenWithBackend(token);

        return token;
      } else {
        console.warn("No se pudo obtener el token FCM");
        return null;
      }
    } catch (error) {
      console.error("Error al obtener token FCM:", error);
      return null;
    }
  }

  /**
   * Configurar listeners para mensajes de Firebase
   */
  private setupMessageListeners(): void {
    // Escuchar mensajes cuando la app está en primer plano
    onMessage(messaging, (payload) => {
      console.log("Mensaje FCM recibido:", payload);

      // Mostrar notificación del navegador
      this.showBrowserNotification(payload);

      // Callback personalizado
      if (this.onMessageCallback) {
        this.onMessageCallback(payload);
      }

      // Marcar como entregada en el backend
      this.markNotificationAsDelivered(payload);
    });
  }

  /**
   * Mostrar notificación del navegador
   */
  private async showBrowserNotification(payload: any): Promise<void> {
    if (!("Notification" in window) || Notification.permission !== "granted") {
      return;
    }

    const title = payload.notification?.title || "Nueva notificación";
    const options: NotificationOptions = {
      body: payload.notification?.body || "",
      icon: "/movifleet.svg", // Tu icono de app
      badge: "/badge-72x72.png", // Badge para la notificación
      tag: payload.data?.tipo || "general", // Tag para agrupar notificaciones
      data: payload.data, // Datos adicionales
      requireInteraction: payload.data?.prioridad === "critica", // Requiere interacción si es crítica
    };

    const notification = new Notification(title, options);

    // Manejar clic en la notificación
    notification.onclick = (event) => {
      event.preventDefault();
      notification.close();

      // Enfocar la ventana
      if (window) {
        window.focus();
      }

      // Navegar según el tipo de notificación
      this.handleNotificationClick(payload);

      // Marcar como leída
      this.markNotificationAsRead(payload);
    };

    // Cerrar automáticamente después de un tiempo (opcional)
    setTimeout(() => {
      notification.close();
    }, 10000); // 10 segundos
  }

  /**
   * Registrar token FCM en el backend
   */
  private async registerTokenWithBackend(token: string): Promise<void> {
    try {
      const authToken = localStorage.getItem("access_token");
      if (!authToken) {
        console.warn("No hay token de autenticación");
        return;
      }

      const baseUrl = getApiBaseUrl();
      const response = await fetch(
        `${baseUrl}/api/notificaciones/dispositivofcm/`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${authToken}`,
          },
          body: JSON.stringify({
            token: token,
            tipo_dispositivo: "web",
            nombre_dispositivo: `${
              navigator.userAgent.includes("Chrome") ? "Chrome" : "Browser"
            } - ${navigator.platform}`,
          }),
        }
      );

      if (response.ok) {
        console.log("✅ Token registrado en el backend");
      } else {
        console.error("❌ Error al registrar token:", response.status);
      }
    } catch (error) {
      console.error("❌ Error al registrar token:", error);
    }
  }

  /**
   * Marcar notificación como entregada
   */
  private async markNotificationAsDelivered(payload: any): Promise<void> {
    const notificationId = payload.data?.notificacion_id;
    if (!notificationId) return;

    try {
      const authToken = localStorage.getItem("access_token");
      if (!authToken) return;

      const baseUrl = getApiBaseUrl();
      await fetch(
        `${baseUrl}/api/notificaciones/notificaciones/marcar_multiples/`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${authToken}`,
          },
          body: JSON.stringify({
            notificacion_ids: [parseInt(notificationId)],
            accion: "marcar_entregada",
          }),
        }
      );
    } catch (error) {
      console.error("Error al marcar como entregada:", error);
    }
  }

  /**
   * Marcar notificación como leída
   */
  private async markNotificationAsRead(payload: any): Promise<void> {
    const notificationId = payload.data?.notificacion_id;
    if (!notificationId) return;

    try {
      const authToken = localStorage.getItem("access_token");
      if (!authToken) return;

      const baseUrl = getApiBaseUrl();
      await fetch(
        `${baseUrl}/api/notificaciones/notificaciones/${notificationId}/marcar_leida/`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${authToken}`,
          },
        }
      );
    } catch (error) {
      console.error("Error al marcar como leída:", error);
    }
  }

  /**
   * Manejar clic en notificación para navegación
   */
  private handleNotificationClick(payload: any): void {
    const tipo = payload.data?.tipo;
    const viajeId = payload.data?.viaje_id;

    console.log("🧭 Navegando por notificación tipo:", tipo);

    // Aquí implementarías la navegación usando tu router
    switch (tipo) {
      case "nuevo_viaje":
        // Navegar a lista de viajes disponibles
        window.location.href = "/client/viajes-disponibles";
        break;
      case "viaje_cancelado":
        // Navegar a mis reservas o detalles del viaje
        if (viajeId) {
          window.location.href = `/client/viaje/${viajeId}`;
        }
        break;
      case "recordatorio":
        // Navegar a detalles del viaje
        if (viajeId) {
          window.location.href = `/client/viaje/${viajeId}`;
        }
        break;
      default:
        // Navegación por defecto
        window.location.href = "/";
    }
  }

  /**
   * Obtener notificaciones del usuario
   */
  public async getNotifications(): Promise<NotificationData[]> {
    try {
      const authToken = localStorage.getItem("access_token");
      if (!authToken) {
        console.warn(
          "🔔 [NotificationService] No hay token de autenticación disponible"
        );
        alert(
          "No tienes sesión activa. Por favor, inicia sesión para ver tus notificaciones."
        );
        return [];
      }

      const baseUrl = getApiBaseUrl();
      const url = `${baseUrl}/api/notificaciones/notificaciones/`;

      const response = await fetch(url, {
        headers: {
          Authorization: `Bearer ${authToken}`,
          "Content-Type": "application/json",
        },
      });


      if (!response.ok) {
        if (response.status === 401) {
          alert(
            "Token de autenticación inválido o expirado. Por favor, inicia sesión nuevamente."
          );
          localStorage.removeItem("access_token");
        }
        const errorText = await response.text();
        console.error("🔔 [NotificationService] Error response:", errorText);
        throw new Error(
          `Error HTTP ${response.status}: ${response.statusText}`
        );
      }

      const data = await response.json();

      const notifications = data.results || data;

      if (!Array.isArray(notifications)) {
        return [];
      }

      // Computar el campo fue_leida basado en fecha_lectura
      const processedNotifications = notifications.map((notification: any) => ({
        ...notification,
        id: String(notification.id), // Asegurar que el ID sea string
        fue_leida:
          notification.fecha_lectura !== null &&
          notification.fecha_lectura !== undefined,
      }));

      return processedNotifications;
    } catch (error) {
      console.error(
        "🔔 [NotificationService] Error al obtener notificaciones:",
        error
      );
      return [];
    }
  }

  /**
   * Obtener preferencias de notificación
   */
  public async getPreferences(): Promise<NotificationPreferences | null> {
    try {
      const authToken = localStorage.getItem("access_token");
      if (!authToken) throw new Error("No autenticado");

      const baseUrl = getApiBaseUrl();
      const response = await fetch(
        `${baseUrl}/api/notificaciones/preferencias/`,
        {
          headers: {
            Authorization: `Bearer ${authToken}`,
          },
        }
      );

      if (!response.ok) throw new Error("Error al obtener preferencias");

      return await response.json();
    } catch (error) {
      console.error("Error al obtener preferencias:", error);
      return null;
    }
  }

  /**
   * Actualizar preferencias de notificación
   */
  public async updatePreferences(
    preferences: Partial<NotificationPreferences>
  ): Promise<boolean> {
    try {
      const authToken = localStorage.getItem("authToken");
      if (!authToken) throw new Error("No autenticado");

      const baseUrl = getApiBaseUrl();
      const response = await fetch(
        `${baseUrl}/api/notificaciones/preferencias/`,
        {
          method: "PUT",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${authToken}`,
          },
          body: JSON.stringify(preferences),
        }
      );

      return response.ok;
    } catch (error) {
      console.error("Error al actualizar preferencias:", error);
      return false;
    }
  }

  /**
   * Configurar callback para mensajes recibidos
   */
  public onMessage(callback: (payload: any) => void): void {
    this.onMessageCallback = callback;
  }

  /**
   * Obtener token actual
   */
  public getCurrentToken(): string | null {
    return this.currentToken;
  }

  /**
   * Verificar si las notificaciones están habilitadas
   */
  public areNotificationsEnabled(): boolean {
    return Notification.permission === "granted";
  }

  /**
   * Desregistrar token del backend
   */
  public async unregisterToken(): Promise<void> {
    try {
      // Implementar lógica para desregistrar el token del backend
      console.log("🚫 Desregistrando token del backend");
      this.currentToken = null;
    } catch (error) {
      console.error("Error al desregistrar token:", error);
    }
  }
}

// Exportar singleton
export default NotificationService.getInstance();
